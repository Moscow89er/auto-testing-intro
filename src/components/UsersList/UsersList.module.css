.usersList {
  padding: 20px;
  margin: 0 auto;
  max-width: 800px;
  background-color: #fff;
}

.user {
  border: 1px solid #ccc;
  border-radius: 8px;
  padding: 15px;
  margin-bottom: 10px;
  background-color: #f2f2f2;
}

.user h3 {
  margin: 0 0 5px;
  font-size: 1.2em;
}

.user p {
  margin: 0;
  color: #555;
}
