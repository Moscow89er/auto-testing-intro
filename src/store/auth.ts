import { createSlice } from "@reduxjs/toolkit";

type InitialAuthState = {
  isAuthenticated: boolean;
};

const initialAuthState: InitialAuthState = {
  isAuthenticated: false,
};

const authSlice = createSlice({
  name: "authentication",
  initialState: initialAuthState,
  reducers: {
    login(state) {
      state.isAuthenticated = true;
    },
    logout(state) {
      state.isAuthenticated = false;
    },
  },
});

export const authActions = authSlice.actions;

export default authSlice.reducer;
