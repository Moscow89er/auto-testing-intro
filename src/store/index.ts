import { combineReducers, configureStore } from "@reduxjs/toolkit";
import type { TypedUseSelectorHook } from "react-redux";
import { useDispatch, useSelector } from "react-redux";
import authReducer from "./auth";
import counterReducer from "./counter";
import userReducer from "./user";

// const rootReducer = { counter: authSlice, auth: counterSlice };

const rootReducer = combineReducers({
  authReducer,
  counterReducer,
  userReducer,
});

const store = configureStore({
  reducer: rootReducer,
});

export type RootState = ReturnType<typeof store.getState>; // получаем тип состояния
export type AppDispatch = typeof store.dispatch; // позволяет использовать только те dispatch, которые мы получили

// хуки для работы с типизированными данными
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export default store;
