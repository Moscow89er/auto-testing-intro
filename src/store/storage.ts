export const createStorage = () => {
    let kvStorage: Record<string, string> | null = null;

    return {
        // инициализация хранилища
        init: (): void => {
            kvStorage = {};
        },
        // очищение хранилища
        clear: (): void => {
            kvStorage = null;
        },
        // добавление элемента
        add: (key: string, value: string): void => {
            if (kvStorage !== null) {
                kvStorage[key] = value;
            }
        },
        // удаление элемента
        remove: (key: string): void => {
            if (kvStorage !== null) {
                delete kvStorage[key];
            }
        },
        // получаем количество элементов в хранилище
        getStorageLength: (): number | null => {
            return kvStorage !== null ? Object.keys(kvStorage).length : null;
        }
    };
};
