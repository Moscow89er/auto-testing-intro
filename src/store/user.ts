import { createSlice, PayloadAction } from "@reduxjs/toolkit";

type User = {
  id: number;
  name: string;
  email: string;
  isLiked: boolean;
};

type UserState = {
  users: User[];
  isLoading: boolean;
  error: string;
};

const initialState: UserState = {
  users: [],
  isLoading: false,
  error: "",
};

export const userSlice = createSlice({
  name: "user",
  initialState,
  reducers: {
    usersFetching(state) {
      state.isLoading = true;
    },
    usersFetchingSuccess(state, action: PayloadAction<Omit<User, "likes">[]>) {
      state.isLoading = false;
      state.error = "";
      state.users = action.payload.map((user) => ({ ...user, likes: 0 }));
    },
    usersFetchingError(state, action: PayloadAction<string>) {
      state.isLoading = false;
      state.error = action.payload;
    },
    toggleLike(state, action: PayloadAction<number>) {
      const user = state.users.find((u) => u.id === action.payload);
      if (user) {
        user.isLiked = !user.isLiked;
      }
    },
  },
});

export default userSlice.reducer;
export const { toggleLike } = userSlice.actions;
// userSlice.ts (добавляем селектор)
export const selectLikedUsers = (state: { userReducer: UserState }) =>
  state.userReducer.users.filter((user) => user.isLiked);
