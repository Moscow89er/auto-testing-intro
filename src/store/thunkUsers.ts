import { AppDispatch } from "../store";
import { userSlice } from "./user";

export const fetchUsers = () => async (dispatch: AppDispatch) => {
  try {
    dispatch(userSlice.actions.usersFetching());
    const response = await fetch("https://jsonplaceholder.typicode.com/users");
    const users = await response.json();
    dispatch(userSlice.actions.usersFetchingSuccess(users));
  } catch (error) {
    if (error instanceof Error) {
      dispatch(userSlice.actions.usersFetchingError(error.message));
    } else {
      dispatch(userSlice.actions.usersFetchingError("Неизвестная ошибка"));
    }
  }
};
