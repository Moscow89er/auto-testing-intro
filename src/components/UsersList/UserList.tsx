import { FC, useEffect } from "react";
import { useAppDispatch, useAppSelector } from "../../store";
import { fetchUsers } from "../../store/thunkUsers";
import { userSlice } from "../../store/user";
import classes from "./UsersList.module.css";

const UsersList: FC = () => {
  const dispatch = useAppDispatch();
  const { users, isLoading, error } = useAppSelector(
    (state) => state.userReducer,
  );

  useEffect(() => {
    dispatch(fetchUsers());
  }, [dispatch]);

  return (
    <main className={classes.usersList}>
      <h2>Список друзей</h2>
      {isLoading && <h1>Идет загрузка...</h1>}
      {error && <h1>{error}</h1>}
      <div>
        {users.map((user) => (
          <div key={user.id} className={classes.user}>
            <h3>{user.name}</h3>
            <p>{user.email}</p>
            <button
              className={classes.button}
              onClick={() => dispatch(userSlice.actions.toggleLike(user.id))}
            >
              {user.isLiked
                ? "Лайк (кликни, чтобы убрать)"
                : "Кликни, чтобы поставить лайк"}
            </button>
          </div>
        ))}
      </div>
    </main>
  );
};

export default UsersList;
