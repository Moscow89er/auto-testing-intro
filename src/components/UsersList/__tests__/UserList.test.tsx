import { vi } from "vitest";
import userReducer, { selectLikedUsers, toggleLike } from "../../../store/user";
import { configureStore } from "@reduxjs/toolkit";
import { fetchUsers } from "../../../store/thunkUsers";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import UsersList from "../UserList";

describe("Тесты синхронных экшенов userSlice", () => {
  // Начальное состояние, которое будем изменять в тестах
  const initialUserState = {
    users: [
      {
        id: 1,
        name: "John Doe",
        email: "john@example.com",
        isLiked: false,
      },
      {
        id: 2,
        name: "Jane Doe",
        email: "jane@example.com",
        isLiked: true,
      },
    ],
    isLoading: false,
    error: "",
  };

  test("поставить лайк пользователю", () => {
    // Применяем экшен toggleLike для пользователя с id: 1
    const newState = userReducer(initialUserState, toggleLike(1));
    // Проверяем, что состояние изменилось ожидаемым образом
    expect(newState.users).toEqual([
      {
        id: 1,
        name: "John Doe",
        email: "john@example.com",
        isLiked: true,
      },
      {
        id: 2,
        name: "Jane Doe",
        email: "jane@example.com",
        isLiked: true,
      },
    ]);
  });

  test("снять лайк у пользователя", () => {
    // Применяем экшен toggleLike для пользователя с id: 2
    const newState = userReducer(initialUserState, toggleLike(2));
    // Проверяем, что состояние изменилось ожидаемым образом
    expect(newState.users).toEqual([
      {
        id: 1,
        name: "John Doe",
        email: "john@example.com",
        isLiked: false,
      },
      {
        id: 2,
        name: "Jane Doe",
        email: "jane@example.com",
        isLiked: false,
      },
    ]);
  });
});

describe("Тест асинхронных экшенов userSlice", () => {
  it("тест загрузки пользователей", async () => {
    const expectedResult = [
      {
        id: 1,
        name: "John Doe",
        email: "john@example.com",
        isLiked: false,
      },
      {
        id: 2,
        name: "Jane Doe",
        email: "jane@example.com",
        isLiked: true,
      },
    ];

    // Мокаем fetch так, чтобы он возвращал ожидаемые данные
    global.fetch = vi.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve(expectedResult),
      }),
    ) as unknown as typeof fetch;

    // Создаем стор, в который будут помещаться данные, полученные через fetchUsers
    const store = configureStore({ reducer: { userReducer } });
    type AppDispatch = typeof store.dispatch;
    // Ожидаем выполнения асинхронного экшена fetchUsers
    await store.dispatch(fetchUsers() as AppDispatch);
    // Получаем состояние стора
    const { users } = store.getState().userReducer;

    // Проверяем, что данные в сторе соответствуют ожидаемому результату с добавленным полем likes: 0
    expect(users).toEqual(
      expectedResult.map((user) => ({ ...user, likes: 0 })),
    );
  });
});

describe("Тесты селекторов userSlice", () => {
  it("получение лайкнутых пользователей", () => {
    // Создаем стор с предустановленным состоянием, где 2 из 3 пользователей имеют лайк
    const store = configureStore({
      reducer: {
        userReducer: userReducer,
      },
      preloadedState: {
        userReducer: {
          users: [
            {
              id: 1,
              name: "John Doe",
              email: "john@example.com",
              isLiked: false,
              likes: 0,
            },
            {
              id: 2,
              name: "Jane Doe",
              email: "jane@example.com",
              isLiked: true,
              likes: 0,
            },
            {
              id: 3,
              name: "Alex Smith",
              email: "alex@example.com",
              isLiked: true,
              likes: 0,
            },
          ],
          isLoading: false,
          error: "",
        },
      },
    });

    // Получаем лайкнутых пользователей с помощью селектора
    const likedUsers = selectLikedUsers(store.getState());

    // Проверяем, что селектор возвращает только пользователей с isLiked === true
    expect(likedUsers).toEqual([
      {
        id: 2,
        name: "Jane Doe",
        email: "jane@example.com",
        isLiked: true,
        likes: 0,
      },
      {
        id: 3,
        name: "Alex Smith",
        email: "alex@example.com",
        isLiked: true,
        likes: 0,
      },
    ]);
  });
});

describe("Снапшот-тест компонента UsersList", () => {
  it("отображается корректно и соответствует сохранённому снапшоту", () => {
    // Задаем предустановленное состояние для редьюсера userReducer
    const preloadedState = {
      userReducer: {
        users: [
          {
            id: 1,
            name: "John Doe",
            email: "john@example.com",
            isLiked: false,
            likes: 0,
          },
          {
            id: 2,
            name: "Jane Doe",
            email: "jane@example.com",
            isLiked: true,
            likes: 0,
          },
        ],
        isLoading: false,
        error: "",
      },
    };

    // Создаем тестовый стор с помощью configureStore
    const store = configureStore({
      reducer: { userReducer },
      preloadedState,
    });

    // Рендерим компонент, оборачивая его в Provider с нашим тестовым стором
    const { asFragment } = render(
      <Provider store={store}>
        <UsersList />
      </Provider>,
    );

    // Сравниваем сгенерированный фрагмент DOM с сохраненным снапшотом
    expect(asFragment()).toMatchSnapshot();
  });
});

// Мокаем асинхронный экшен fetchUsers, чтобы он не перезаписывал наше предустановленное состояние в тесте
vi.mock("../../store/thunkUsers", () => ({
  fetchUsers: () => () => {},
}));

describe("Тест обработчика клика в компоненте UsersList", () => {
  it("должен переключать статус лайка при клике на кнопку", () => {
    // Задаем предустановленное состояние: один пользователь без лайка
    const preloadedState = {
      userReducer: {
        users: [
          {
            id: 1,
            name: "John Doe",
            email: "john@example.com",
            isLiked: false,
            likes: 0,
          },
        ],
        isLoading: false,
        error: "",
      },
    };

    // Создаем тестовый стор с предустановленным состоянием
    const store = configureStore({
      reducer: { userReducer },
      preloadedState,
    });

    // Рендерим компонент, оборачивая его в Provider с нашим тестовым стором
    render(
      <Provider store={store}>
        <UsersList />
      </Provider>,
    );

    // Находим кнопку по её тексту, который отображается для не лайкнутого пользователя
    const likeButton = screen.getByRole("button", {
      name: /Кликни, чтобы поставить лайк/i,
    });
    expect(likeButton).toBeInTheDocument();

    // Имитируем клик по кнопке
    fireEvent.click(likeButton);

    // После клика текст кнопки должен измениться, так как статус лайка изменился
    expect(
      screen.getByRole("button", { name: /Лайк \(кликни, чтобы убрать\)/i }),
    ).toBeInTheDocument();
  });
});
