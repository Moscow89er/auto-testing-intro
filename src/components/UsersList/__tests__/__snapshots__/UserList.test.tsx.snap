// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Снапшот-тест компонента UsersList > отображается корректно и соответствует сохранённому снапшоту 1`] = `
<DocumentFragment>
  <main
    class="_usersList_709914"
  >
    <h2>
      Список друзей
    </h2>
    <h1>
      Идет загрузка...
    </h1>
    <div>
      <div
        class="_user_709914"
      >
        <h3>
          John Doe
        </h3>
        <p>
          john@example.com
        </p>
        <button
          class="_button_709914"
        >
          Кликни, чтобы поставить лайк
        </button>
      </div>
      <div
        class="_user_709914"
      >
        <h3>
          Jane Doe
        </h3>
        <p>
          jane@example.com
        </p>
        <button
          class="_button_709914"
        >
          Лайк (кликни, чтобы убрать)
        </button>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
